### PR TITLE
[Quantization][Pytorch] On device quantization support part 1

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84131
* #83807
* #83742
* #83571
* #83570
* #83569
* __->__ #83568

Summary:
TO support on device quantization this diff introduces observer
insertion. Specifically observers are inserted by adding new method with
prefix observ_.

Intent is that post training, this method will be run to record
statistics

Test Plan:
test_ondevice_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D38771417](https://our.internmc.facebook.com/intern/diff/D38771417)

### CODE REVIEW

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

File: torch/ao/quantization/quantize_jit.py [hunk 1]
```diff
@@ -62,6 +62,25 @@ def _prepare_jit(model, qconfig_dict, inplace=False, quant_type=QuantType.STATIC
...
 
+def _prepare_ondevice_jit(model, qconfig_dict, method_name='forward', inplace=False, quant_type=QuantType.STATIC):
+    _check_is_script_module(model)
+    if not all(isinstance(x, str) for x in qconfig_dict.keys()):
+        raise ValueError('qconfig_dict should only contain names(str) as keys.')
+    scripted_qconfig_dict = script_qconfig_dict(qconfig_dict)
+    method_graph = model._c._get_method(method_name).graph
+    torch._C._jit_pass_inline(method_graph)
+    model = fuse_conv_bn_jit(model, inplace)
+    model_c = torch._C._jit_pass_insert_observer_method_for_ondevice_ptq(model._c,
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize_jit.py:73]: nit: lint

**Review by jerryzh168 — APPROVED** [review]
since quantize_jit is no longer maintained. I think we are OK as long as this works for you

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

File: test/allowlist_for_publicAPI.json [hunk 2]
```diff
@@ -1469,6 +1469,7 @@
     "no_observer_set",
     "prepare",
     "prepare_dynamic_jit",
+    "prepare_ondevice_dynamic_jit",
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:1472]: Nothing should be added to these lists.
You should make sure to fix the corresponding submodule.

**Review by albanD — DISMISSED** [review]

File: test/allowlist_for_publicAPI.json (hunk 2 above)
Comment (kimishpatel) [reply test/allowlist_for_publicAPI.json:1472]: @albanD Meaning? Add `__moduel__` tag to the API?

Comment (kimishpatel) [reply test/allowlist_for_publicAPI.json:1472]: I will actually just prefix this with `_` since I dont envision us having to support this for long term.

Comment (albanD) [reply test/allowlist_for_publicAPI.json:1472]: if they should just be private, prefixing with `_` will work indeed.

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38771417).

**Pull request closed without merging**